File: opensef/opensef-appkit/src/OSFView.m
/**
 * OSFView.m - Base view implementation
 * openSEF Framework
 */

#import <opensef/OpenSEFAppKit.h>

@implementation OSFView

- (instancetype)init {
  self = [super init];
  if (self) {
    _frame = CGRectZero;
    _subviews = [NSMutableArray array];
    _alpha = 1.0;
    _hidden = NO;
  }
  return self;
}

- (void)addSubview:(OSFView *)view {
  view.superview = self;
  [self.subviews addObject:view];
}

- (void)removeFromSuperview {
  [self.superview.subviews removeObject:self];
  self.superview = nil;
}

- (void)draw:(CGContextRef)context {
  // Override in subclasses for custom drawing
  for (OSFView *subview in self.subviews) {
    if (!subview.hidden) {
      [subview draw:context];
    }
  }
}

@end

File: opensef/opensef-appkit/src/OSFView.cpp
/**
 * OSFView.cpp - Base View Implementation
 */

#include <algorithm>
#include <opensef/OpenSEFAppKit.h>


namespace opensef {

void OSFView::addSubview(std::shared_ptr<OSFView> view) {
  if (view->superview_) {
    view->removeFromSuperview();
  }
  view->superview_ = this;
  subviews_.push_back(view);
}

void OSFView::removeFromSuperview() {
  if (!superview_)
    return;

  auto &siblings = superview_->subviews_;
  siblings.erase(std::remove_if(siblings.begin(), siblings.end(),
                                [this](const std::shared_ptr<OSFView> &v) {
                                  return v.get() == this;
                                }),
                 siblings.end());
  superview_ = nullptr;
}

void OSFView::render(cairo_t *cr) {
  if (hidden_)
    return;

  // Render subviews
  for (auto &subview : subviews_) {
    subview->render(cr);
  }
}

} // namespace opensef
